### PR TITLE
Major API change on JSClosure

### PR DESCRIPTION
**Pull request by kateinoigakukun**
## Motivation

While integrating async/await feature, I found `JSPromise` has too complex API. (e.g. too many overloads, unsafe generic parameters, and lifetime management)
So I want to refine the API before async/await integration. As a first step, I changed `JSClosure` API.

## Changes

- Add `JSOneshotClosure` which provides an ability to invoke a closure at most once. This will be suitable to use with functions like `Promise.then`
- Remove `JSClosure.init(_ body: @escaping ([JSValue]) -> Void)` overload, which forces users to write result type in closure.
- Add JSClosure lifetime test suites

### CODE REVIEW

File: Sources/JavaScriptKit/FundamentalObjects/JSFunction.swift
Comment (MaxDesiatov) [inline Sources/JavaScriptKit/FundamentalObjects/JSFunction.swift:226]: Can we keep this overload for at least one minor version as deprecated? This would give people some time to migrate away from it.

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (MaxDesiatov) [inline Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:5]:
```suggestion
/// `JSOneshotClosure` is a JavaScript function that can be called only once.
```

File: IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift
Comment (MaxDesiatov) [inline IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift:219]:
```suggestion
        // second call will cause `fatalError` that can be caught as a JavaScript exception
```

**Review by MaxDesiatov — COMMENTED** [review]
Overall this seems legit with a few suggestions in comments.

**Review by MaxDesiatov — APPROVED** [review]
I wasn't sure about that `JSClosure.init` overload myself when I introduced, but you're right it brings more harm than good. Also, great idea to simplify a lot of things with `JSOneshotClosure`! 👍

**Comment by kateinoigakukun** [comment]
@swiftwasm/jskit-team Could anyone review this change? I want two or more approvals for this kind of API change before merging  🙏

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (j-f1) [inline Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:123]: I’ve usually seen this referred to as just a “thunk:”

```suggestion
// 1. Create a thunk in the JavaScript world, which has a reference
//    to a Swift closure.
// ┌─────────────────────┬──────────────────────────┐
// │     Swift side      │   JavaScript side        │
// │                     │                          │
// │                     │                          │
// │                     │   ┌──────[Thunk]───────┐ │
// │          ┌ ─ ─ ─ ─ ─│─ ─│─ ─ ─ ─ ─ ┐         │ │
// │          ↓          │   │          │         │ │
// │  [Swift Closure]    │   │  Host Function ID  │ │
// │                     │   │                    │ │
// │                     │   └────────────────────┘ │
// └─────────────────────┴──────────────────────────┘
//
// 2. When the thunk function is invoked, it calls the Swift closure via
//    `_call_host_function` and receives the result through a callback.
// ┌─────────────────────┬──────────────────────────┐
// │     Swift side      │   JavaScript side        │
// │                     │                          │
// │                     │                          │
// │                   Apply ┌──────[Thunk]───────┐ │
// │          ┌ ─ ─ ─ ─ ─│─ ─│─ ─ ─ ─ ─ ┐         │ │
// │          ↓          │   │          │         │ │
// │  [Swift Closure]    │   │  Host Function ID  │ │
// │          │          │   │                    │ │
// │          │          │   └────────────────────┘ │
// │          │          │                    ↑     │
// │          │        Apply                  │     │
// │          └─[Result]─┼───>[Callback func]─┘     │
// │                     │                          │
// └─────────────────────┴──────────────────────────┘
```

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (MaxDesiatov) [reply Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:66]: The issue is with Swift typechecker which can't treat the currently available set of `JSClosure.init` overloads unambiguously without adding an explicit function signature at the place of use ([example one](https://github.com/TokamakUI/Tokamak/blob/a631d181e69e2e72d898b0b778356e61fb03e93b/Sources/TokamakDOM/App/ScenePhaseObserver.swift#L24),  [example two](https://github.com/TokamakUI/Tokamak/blob/3451d9ea12d28f060aa9fa84547e4fdacfc63a8d/Sources/TokamakTestRenderer/TestRenderer.swift#L27)). That defeats the purpose of the overload. I think we'd rather require users to add `return .undefined` explicitly (which can be easily copy-pasted where needed) instead of adding function signatures explicitly (which are almost always ad hoc).

File: Sources/JavaScriptKit/BasicObjects/JSTimer.swift
Comment (j-f1) [inline Sources/JavaScriptKit/BasicObjects/JSTimer.swift:37]: If `isRepeating` is false, this can be a one-shot closure, right?

File: Sources/JavaScriptKit/BasicObjects/JSPromise.swift
Comment (j-f1) [inline Sources/JavaScriptKit/BasicObjects/JSPromise.swift:55]: All promise closures can be one-shot, so I’m pretty sure we don’t need to hold on to them manually on the Swift side.

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (j-f1) [inline Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:27]: I’m wondering — if the function will be destroyed as soon as it’s called, does it even make sense to call a one-shot closure from Swift?

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (j-f1) [inline Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:59]:
```suggestion
    @available(*, deprecated, message: "This initializer will be removed in the next minor version update. Please use `init(_ body: @escaping ([JSValue]) -> JSValue)` and add `return .undefined` to the end of your closure")
```

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (j-f1) [inline Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:86]: I think this makes more sense as a comment
```suggestion
            // Safari doesn't support `FinalizationRegistry`, so we cannot automatically manage the lifetime of Swift objects
            fatalError("release() must be called on JSClosure objects manually before they are deallocated")
```

**Review by j-f1 — COMMENTED** [review]
at a desktop now. LGTM except for these things

File: Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift
Comment (kateinoigakukun) [reply Sources/JavaScriptKit/FundamentalObjects/JSClosure.swift:27]: Yes, usually users don't call `JSOneshotClosure` and `JSClosure` directly from Swift.
It might make sense to remove `JSFunction` inheritance and remove `callAsFunction`.

**Comment by kateinoigakukun** [comment]
@j-f1 I addressed reviewed points. Could you take a look again?

File: Sources/JavaScriptKit/JSValue.swift
Comment (j-f1) [inline Sources/JavaScriptKit/JSValue.swift:148]:
```suggestion
    @available(*, deprecated, renamed: "object", message: "JSClosure is no longer a subclass of JSFunction. Use .object(closure) instead.")
```

File: IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift
Comment (j-f1) [inline IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift:509]: Totally happy to punt this to another issue, but since there isn’t necessarily a need to cancel  a `setTimeout`, could there be an API that doesn’t require holding the `JSTimer` instance? I guess users could just do 

```swift
JSObject.global.setTimeout(JSOneshotClosure {
  // ...
  return undefined
}, 1000)
```

Comment (kateinoigakukun) [reply IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift:509]: Yes, we can provide such API in this signature for example.

```swift
JSTimer.timeout(millisecondsDelay: 100) {
    // ...
}
```

This PR is already too large, so let's implement it in another PR 👍

**Review by jed-fox-neeva — APPROVED** [review]

**Review by j-f1 — APPROVED** [review]
lol oops accounts are hard

**Pull request merged**
